kernel/src/fs/manager.rs

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::sync::{Arc, Weak};
+
+use aster_systree::{
+    impl_cast_methods_for_branch, Error, Result, SysAttrSet, SysBranchNode, SysBranchNodeFields,
+    SysNode, SysNodeId, SysNodeType, SysObj, SysPerms, SysStr,
+};
+use inherit_methods_macro::inherit_methods;
+use ostd::mm::{VmReader, VmWriter};
+use spin::Once;
+
+/// Returns a reference to the [`FsManager`]. Panics if not initialized.
+pub fn singleton() -> &'static Arc<FsManager> {
+    MANAGER.get().expect("Fs manager has not been initialized")
+}
+
+/// Initializes the [`FsManager`] singleton
+pub fn init() {
+    MANAGER.call_once(FsManager::new);
+    aster_systree::singleton()
+        .root()
+        .add_child(singleton().clone())
+        .expect("Failed to add fs manager to SysTree");
+}
+
+static MANAGER: Once<Arc<FsManager>> = Once::new();
+
+/// Manager for virtual filesystem.
+///
+/// Represents the root of all filesystem-related sysfs entries, where:
+/// - Each child node corresponds to a filesystem type (ext2, cgroup, etc.).
+/// - Manages registration and lifetime of filesystem control interfaces.
+#[derive(Debug)]
+pub struct FsManager {
+    fields: SysBranchNodeFields<dyn FsFactory>,
+    weak_self: Weak<Self>,
+}
+
+/// Trait representing a filesystem's control interface.
+pub trait FsFactory: SysObj {
```

Comment:
If you think that this is some kind of control interface, rename the trait to `FsControl`.

I suggested the name `FsFactory` because I believe that the `mount` system call requires each file system to register a factory to create new mount of a particular file system type. But it takes more than a factory to support the functionality of `/proc/filesystems` and `/sys/fs`. So maybe `FsControl` is indeed a better name. As said in a previous comment, the design decisions about the FS registration mechanism is beyond the scope of this PR.